lib/streamlit/widgets.py

```diff
@@ -158,3 +172,21 @@ def _get_widget_value(state: Optional[WidgetState]) -> Optional[Any]:
         return json.loads(getattr(state, value_type))
 
     return getattr(state, value_type)
+
+
+def beta_widget_value(key: str) -> Any:
+    """Returns the value of a widget with a given id, for use in
+    state update callbacks.
+    """
+    import streamlit.report_thread as ReportThread
+    from streamlit.server.server import Server
+    import streamlit.elements.utils as utils
+
+    ctx = ReportThread.get_report_ctx()
+
+    if ctx is None:
+        return None
+
+    this_session = Server.get_current().get_session_by_id(ctx.session_id)
+    this_widget_state = this_session.get_widget_states()
+    return this_widget_state.get_widget_value(utils._get_widget_id("", None, key))
```

Comment:
Looks like we need to deserialize too. Tried this in the PR preview.

```
import streamlit as st

state = st.beta_session_state(s=None)

def checkbox(value):
    state.s = st.beta_widget_value("foo")

vals = ["One", "Two", "Three", "Four"]

value = st.select_slider("Number", vals, key="foo")

st.checkbox("Check me", on_change=checkbox)

st.write(value)
st.write(state.s)
```

Shouldn't be too hard but to get the deserializer in the widget state and run it if possible, though I might move that into the Widget State Manager

Reply:
Had trouble parsing your comment but after running the code locally I understand, and have a rough idea how to fix it. 

Definitely not in scope for this prototype but having a clear marker of serialized vs deserialized values would be very good; possibly as some wrapper class.

Reply:
Ah yes. It seems like, it should be in scope for the prototype, right? Essentially, we want `get_widget_value` to return the same value as the return type from the widget itself otherwise, it would seem confusing.

I believe this method in the class will do the trick and just calling it here instead of `get_widget_value`

```python
def get_deserialized_value(self, widget_id: str) -> Optional[Any]:
        raw_widget_value = self.get_widget_value(widget_id)
        deserializer = self._widget_deserializers.get(widget_id, None)
        if deserializer is None:
            return raw_widget_value
        
        return deserializer(raw_widget_value)
```